# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/IDBCursor.h
# Commit: c009bbdbbfd5
# Full Hash: c009bbdbbfd5451756cb1f3dc421b82df50911d1
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-11-05 22:07:34
# Regressor Bug: 1168606
# File Overlap Count: 1
# Description:
#   Bug 1168606 - Support preloading also for index cursors. r=ttung,asuth
#   
#   Depends on D43252
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D43461
# ==============================================================================

diff -r 89817a19f3b3 -r c009bbdbbfd5 dom/indexedDB/IDBCursor.h
--- a/dom/indexedDB/IDBCursor.h	Tue Nov 05 14:40:37 2019 +0000
+++ b/dom/indexedDB/IDBCursor.h	Tue Nov 05 14:40:35 2019 +0000
@@ -33,6 +33,9 @@
 class BackgroundCursorChild;
 }
 
+// TODO: Consider defining different subclasses for the different cursor types,
+// possibly using the CRTP, which would remove the need for various case
+// distinctions.
 class IDBCursor final : public nsISupports, public nsWrapperCache {
  public:
   typedef indexedDB::Key Key;
@@ -48,7 +51,6 @@
     DIRECTION_INVALID
   };
 
- private:
   enum Type {
     Type_ObjectStore,
     Type_ObjectStoreKey,
@@ -56,6 +58,7 @@
     Type_IndexKey,
   };
 
+ private:
   indexedDB::BackgroundCursorChild* mBackgroundActor;
 
   // TODO: mRequest, mSourceObjectStore and mSourceIndex could be made const if
@@ -74,8 +77,8 @@
   JS::Heap<JS::Value> mCachedValue;
 
   Key mKey;
-  Key mSortKey;
-  Key mPrimaryKey;
+  Key mSortKey;     ///< AKA locale aware key/position elsewhere
+  Key mPrimaryKey;  ///< AKA object store key/position elsewhere
   StructuredCloneReadInfo mCloneInfo;
 
   const Type mType;
@@ -120,6 +123,8 @@
 
   IDBCursorDirection GetDirection() const;
 
+  Type GetType() const;
+
   bool IsContinueCalled() const { return mContinueCalled; }
 
   void GetKey(JSContext* aCx, JS::MutableHandle<JS::Value> aResult,
@@ -163,6 +168,9 @@
 
   void InvalidateCachedResponses();
 
+  // Checks if this is a locale aware cursor (ie. the index's sortKey is unset)
+  bool IsLocaleAware() const;
+
   NS_DECL_CYCLE_COLLECTING_ISUPPORTS
   NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_CLASS(IDBCursor)
 
@@ -176,9 +184,6 @@
 
   ~IDBCursor();
 
-  // Checks if this is a locale aware cursor (ie. the index's sortKey is unset)
-  bool IsLocaleAware() const;
-
   void DropJSObjects();
 
   bool IsSourceDeleted() const;